packages/react-devtools-scheduling-profiler/src/view-base/layouter.js

```diff
@@ -0,0 +1,269 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {Rect} from './geometry';
+import type {View} from './View';
+
+export type LayoutInfo = {|view: View, frame: Rect|};
+export type Layout = LayoutInfo[];
+
+/**
+ * A function that takes a list of subviews, currently laid out in
+ * `existingLayout`, and lays them out into `containingFrame`.
+ */
+export type Layouter = (
+  existingLayout: Layout,
+  containingFrame: Rect,
+) => Layout;
+
+function viewToLayoutInfo(view: View): LayoutInfo {
+  return {view, frame: view.frame};
```

Comment:
> This conversion seems not very useful. Can we get rid of it?

I don't think we can 😆

Because the layouters are composable (via `createComposedLayout`), they need some way to pass the intermediate frames to each other. If we get rid of this `LayoutInfo` wrapper and store the intermediate frames in the views, we have these problems:

1. Invalidated views are redrawn the next time `CanvasPage`'s layout effects are flushed. Because the `View` class's `setFrame` method invalidates the view's contents (by calling `setsNeedsDisplay`) if the frame is changed, we only want to call `setFrame` with the final frame, so that the view will only be redrawn if the frame actually changed.

	Without the `LayoutInfo` wrapper, the intermediate frames will trigger a lot of unnecessary invalidation and thus many views will be unnecessarily redrawn on every frame.

1. View subclasses can override the `setFrame` method and do arbitrary things, including updating its state. For example, `HorizontalPanAndZoomView` and `VerticalScrollView` re-clamp their offsets and zoom levels when the frame changes, so that e.g. resizing the browser window don't cause black areas to appear. If this is done with intermediate frames, this clamping can cause bugs, e.g. if an intermediate frame had a much wider `width` than the final one, resizing the browser window may cause you to suddenly scroll left as the offset got clamped with that intermediate `width`.

This `LayoutInfo` wrapper was the best way I could think of to solve this. Is there a better way to solve this?